Make Tibber priceInfo available as attributes of current price

Knodd opened the pull request:
Added priceInfo for today with timestamp and price level as list in attribute raw_today
Added priceInfo for tomorrow with timestamp and price level as list in attribute raw_tomorrow
Added only prices as list for today and tomorrow as new attributes
Added tomorrow_valid as attribute if array of tomorrows prices are valid
This is the same format as the Nordpool integration, only with Tibber's api keys ("startsAt" and "total")

Adjusted valid update interval by changing hours until last timestamp from 5 to 8 and spread_load_constant from 5000 to 3600 (to reduce refresh window)
This should move update interval from 16:20-18:00 to 13:30-14:30 to make sure tomorrows prices are downloaded a bit earlier in the day

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->



## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
add extra attributes for priceInfo and update them when current price is updated.
The price info is already included in downloaded data

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Knodd committed:
Update sensor.py

Added already downloaded priceInfo for today with timestamp and price level as list in attribute raw_today
Added already downloaded priceInfo for tomorrow with timestamp and price level as list in attribute raw_tomorrow
Added only prices as list for today and tomorrow as new attribute
Added tomorrow_valid as attribute if array of tomorrows prices are valid
This is the same format as the Nordpool integration, only with Tibber's api keys ("startsAt" and "total")

Adjusted valid update interval by changing hours until last timestamp from 5 to 8 and spread_load_constant from 5000 to 3600 (to reduce refresh window)
This should move update interval from 16:20-18:00 to 13:30-14:30 to make sure tomorrows prices are downloaded a bit earlier in the day

Knodd committed:
fix check errors

Changed returing [] to None for empty attributes to make it consistent with initial state for sensor

Knodd committed:
Update sensor.py

fix check errors

Danielhiversen commented:
I do not think these are allowed as attributes.
See this discussion regarding support for future values https://github.com/home-assistant/architecture/discussions/667

Knodd committed:
fix check errors

Knodd commented:
> I do not think these are allowed as attributes. See this discussion regarding support for future values [home-assistant/architecture#667](https://github.com/home-assistant/architecture/discussions/667)

Has this been implemented? If not, would it not make sense to add data as attributes now and move it to forecast later when possible?

I made the format exactly like the Nordpool integration (except for the API-keys to make it clear what data is returned) so it would be an easy swap out for Tibber customers since Nordpool API is deprecated

Danielhiversen commented:
> Has this been implemented? If not, would it not make sense to add data as attributes now and move it to forecast later when possible?

No, but independent of that I do not think these are allowed as attributes.

Knodd commented:
Since it's the same data being used for current price sensor and it already has attributes for min/max/avg, price_level etc I thought it would be ok.

How should we add them then? Can we reuse the data for a separate sensor entity called price_forecast or something? The way I read the link above that's not the optimal way either

Knodd committed:
fix inconsistency

Always return null/None for empty price data
Check that todays prices is valid array before iterating

Knodd committed:
fix check error

Knodd commented:
Who makes the decision if future prices are allowed as attributes of the current price sensor or not?

MartinHjelmare commented:
We should wait for the architecture discussion to be resolved before we expose this information.

Knodd committed:
Always return list for price data

Returning None for empty data makes it necessary to check for datatype i Apexcharts etc. Returning empty list keeps graph from crashing if data is not valid list and returns no data points as expected

Knodd commented:
OK, guess we'll just have to be patient then. Any estimate on the timeline for this discussion?

There has been no comments for 4 months on the discussion and it focuses mostly on the problem with existing integrations spreading forecast values over separate entities making automations based on forecast cumbersome. This is not the case for listed data in attributes where we easily can use {% if states('sensor.electricity_price') > average(state_attr('sensor.electricity_price', 'today')) %} to check if the current price is above todays average price etc

Weather forecast integrations generally (met.no as an example) also exposes forecast data as lists in attributes so I don't see whats worse with forecasting electricity prices as attributes.

I understand the need to keep it organized and in common form, but my opinion is that it would be a better solution to expose the data as attributes now (seems to be the general way integrations solve forecasting data today) and move it to a forcasted values model for sensor data when (and if) such a function is added at some point later, along with all other integrations providing forecasted data

stafbulp commented:
I hope to keep this PR alive. Adding the pricing info would make this integration much more useful for automation purposes.

MartinHjelmare commented:
Let's close here until the architecture discussion is resolved.
